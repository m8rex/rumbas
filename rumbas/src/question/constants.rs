use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use numbas::defaults::DEFAULTS;
use numbas::jme::JMEString;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use comparable::Comparable;

/// Macro used to make sure that the ToNumbas & ToRumbas implementation remain up to data
macro_rules! builtin_constants {
    (
        $(#[$outer:meta])*
        pub struct $struct: ident {
            $(
                $(#[$inner:meta])*
                $field: ident: $type: ty: $name: literal: $default: ident
            ),+
        }
    ) => {
        $(
            #[$outer]
        )*
        pub struct $struct {
            $(
                $(#[$inner])*
                pub $field: $type
            ),*
        }
        impl ToNumbas<std::collections::HashMap<String, bool>> for $struct {
            fn to_numbas(&self, _locale: &str) -> std::collections::HashMap<String, bool> {
                let mut builtin = std::collections::HashMap::new();
                $(
                    builtin.insert($name.to_string(), self.$field);
                )*
                builtin
            }
        }
        impl ToRumbas<BuiltinConstants> for numbas::question::constants::BuiltinConstants {
            fn to_rumbas(&self) -> BuiltinConstants {
                BuiltinConstants {
                $(
                    $field: *self.0.get(&$name.to_string()).unwrap_or(&DEFAULTS.$default)
                ),*
                }
            }
        }
    }
}

builtin_constants! {
    #[derive(Input, Overwrite, RumbasCheck, Examples)]
    #[input(name = "BuiltinConstantsInput")]
    #[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq,)]
    /// Specify which builtin constants should be enabled
    pub struct BuiltinConstants {
        /// Whether the constant e is enabled
        e: bool: "e": builtin_constants_e,
        /// Whether the constant pi is enabled
        pi: bool: "pi,\u{03c0}": builtin_constants_pi,
        /// Whether the constant i is enabled-
        i: bool: "i": builtin_constants_i
    }
}

#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "CustomConstantInput")]
#[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq)]
/// A custom constant
pub struct CustomConstant {
    /// The name of the constant
    pub name: String,
    /// The value of the constant
    pub value: JMEString,
    /// The tex code use to display the constant
    pub tex: String,
}

impl ToNumbas<numbas::question::constants::QuestionConstant> for CustomConstant {
    fn to_numbas(&self, locale: &str) -> numbas::question::constants::QuestionConstant {
        numbas::question::constants::QuestionConstant {
            name: self.name.to_numbas(locale),
            value: self.value.to_numbas(locale),
            tex: self.tex.to_numbas(locale),
        }
    }
}

impl ToRumbas<CustomConstant> for numbas::question::constants::QuestionConstant {
    fn to_rumbas(&self) -> CustomConstant {
        CustomConstant {
            name: self.name.to_rumbas(),
            value: self.value.to_rumbas(),
            tex: self.tex.to_rumbas(),
        }
    }
}
