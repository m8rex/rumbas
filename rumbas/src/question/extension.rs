use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::*;
use comparable::Comparable;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use structdoc::StructDoc;

macro_rules! extensions {
    (
        $(
            $(#[$inner:meta])*
            $name: ident: $path: literal
        ),+
    ) => {
            #[derive(Input, Overwrite, RumbasCheck, Examples, StructDoc)]
            #[input(name = "ExtensionsInput")]
            #[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq, Eq)]
            /// Specify which extensions should be enabled
            pub struct Extensions {
                $(
                    $(
                        #[$inner]
                    )*
                    pub $name: bool
                ),*
            }

        impl ToNumbas<Vec<String>> for Extensions {
            type ToNumbasHelper = ();
            fn to_numbas(&self, locale: &str, _data: &Self::ToNumbasHelper) -> Vec<String> {
                let mut extensions = Vec::new();
                $(
                    if self.$name.to_numbas(locale, &()) {
                        extensions.push($path.to_string()); //TODO: Enum in numbas crate?
                    }
                )*
                extensions
            }
        }

        impl Extensions {
            pub fn from(v: &[String]) -> Self {
                Extensions {
                    $(
                        $name: v.contains(&$path.to_string()).to_rumbas()
                    ),*
                }
            }

            pub fn combine(e: Extensions, f: Extensions) -> Extensions {
                Extensions {
                    $(
                    $name: e.$name || f.$name
                    ),*
                }
            }

            pub fn to_paths(&self) -> Vec<String> {
                let numbas_path = std::env::var(crate::NUMBAS_FOLDER_ENV)
                    .expect(&format!("{} to be set", crate::NUMBAS_FOLDER_ENV)[..]);
                let mut paths = Vec::new();
                $(
                    if self.$name {
                        paths.push($path);
                    }
                )*
                paths
                    .into_iter()
                    .map(|s| format!("{}/extensions/{}", numbas_path, s))
                    .collect()
            }
        }

        impl Default for Extensions {
            fn default() -> Extensions {
                Extensions {
                    $(
                        $name: false
                    ),*
                }
            }
        }

    }
}

//fixme vis.js extension? (where to find?)
extensions! {
    chemistry: "chemistry",
    download_text_file: "download-text-file",
    /// The Eukleides extension provides functions to embed diagrams created using the
    /// Eukleides language.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#eukleides
    eukleides: "eukleides",
    /// The GeoGebra extension provides functions to embed GeoGebra worksheets in a question.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#geogebra
    geogebra: "geogebra",
    /// This extension provides some functions for working with and drawing graphs (networks of vertices joined by edges) in Numbas.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#graph-theory
    graphs: "graphs",
    /// The JSXGraph extension provides functions to create and manipulate interactive diagrams with the JSXGraph library.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#eukleides
    jsx_graph: "jsxgraph",
    linear_algebra: "linear-algebra",
    /// This extension provides a new data type and some functions to deal with linear codes.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#linear-codes
    linear_codes: "codewords",
    optimisation: "optimisation",
    permutations: "permutations",
    /// This extension provides a new data type and some functions to deal with rational polynomials.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#polynomials
    polynomials: "polynomials",
    /// This extension wraps the js-quantities library to provide a “quantity with units” data type to Numbas.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#quantities
    quantities: "quantities",
    /// The “random person” extension provides a collection of functions to generate random people, for use in word problems.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#random-person
    random_person: "random_person",
    /// The statistical functions extension provides many new functions for generating samples from random distributions, and calculating statistics.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#statistical-functions
    stats: "stats",
    sqlite: "sqlite",
    text: "text",
    written_number: "written-number"

    // TODO: programming extension
}

impl ToRumbas<Extensions> for Vec<String> {
    fn to_rumbas(&self) -> Extensions {
        Extensions::from(self)
    }
}
