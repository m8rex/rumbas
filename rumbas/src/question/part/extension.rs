use crate::question::part::question_part::JMENotes;
use crate::question::part::question_part::JMENotesInput;
use crate::question::part::question_part::VariableReplacementStrategy;
use crate::question::part::question_part::{QuestionPartInput, VariableReplacementStrategyInput};
use crate::question::QuestionParts;
use crate::question::QuestionPartsInput;
use crate::support::optional_overwrite::*;
use crate::support::rumbas_types::*;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::*;
use crate::support::translatable::ContentAreaTranslatableString;
use crate::support::translatable::ContentAreaTranslatableStringInput;
use numbas::support::primitive::Primitive;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

question_part_type! {
    pub struct QuestionPartExtension {}
}

impl ToNumbas<numbas::question::part::extension::QuestionPartExtension> for QuestionPartExtension {
    fn to_numbas(&self, locale: &str) -> numbas::question::part::extension::QuestionPartExtension {
        numbas::question::part::extension::QuestionPartExtension {
            part_data: self.to_numbas(locale),
        }
    }
}

impl ToRumbas<QuestionPartExtension> for numbas::question::part::extension::QuestionPartExtension {
    fn to_rumbas(&self) -> QuestionPartExtension {
        create_question_part! {QuestionPartExtension with &self.part_data => {}}
    }
}
