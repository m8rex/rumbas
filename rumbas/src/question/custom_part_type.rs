use crate::question::extension::Extensions;
use crate::question::part::question_part::JMENotes;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::JMETranslatableString;
use crate::support::translatable::TranslatableString;
use crate::support::translatable::TranslatableStrings;
use crate::support::yaml::{YamlError, YamlResult};
use numbas::question::custom_part_type::CustomPartTypeSetting as NCustomPartTypeSetting;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::convert::TryInto;
use std::hash::{Hash, Hasher};

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "CustomPartTypeDefinitionInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
pub struct CustomPartTypeDefinition {
    pub type_name: TranslatableString,
    pub description: TranslatableString,
    pub settings: Vec<NCustomPartTypeSetting>, // TODO
    pub can_be_gap: bool,
    pub can_be_step: bool,
    pub marking_notes: JMENotes,
    pub help_url: TranslatableString,
    pub published: bool,
    pub extensions: Extensions,
    pub input_widget: CustomPartInputWidget, //TODO source
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartType> for CustomPartTypeDefinition {
    fn to_numbas(&self, _locale: &str) -> numbas::question::custom_part_type::CustomPartType {
        panic!(
            "{}",
            "Should not happen, don't call this method Missing name".to_string(),
        )
    }
    fn to_numbas_with_name(
        &self,
        locale: &str,
        name: String,
    ) -> numbas::question::custom_part_type::CustomPartType {
        numbas::question::custom_part_type::CustomPartType {
            short_name: name,
            name: self.type_name.clone().to_string(locale).unwrap(),
            description: self.description.clone().to_string(locale).unwrap(),
            settings: self.settings.clone(), // .to_numbas(&locale).unwrap(),
            help_url: self.help_url.clone().to_string(locale).unwrap(),
            public_availability: numbas::question::custom_part_type::CustomPartAvailability::Always,
            marking_script: self.marking_notes.to_numbas(locale),
            can_be_gap: self.can_be_gap,
            can_be_step: self.can_be_step,
            marking_notes: self.marking_notes.0.to_numbas(locale),
            published: self.published,
            extensions: self.extensions.to_numbas(locale),
            input_widget: self.input_widget.to_numbas(locale),
        }
    }
}

impl CustomPartTypeDefinitionInput {
    pub fn from_name(name: &str) -> YamlResult<Self> {
        let file =
            std::path::Path::new(crate::CUSTOM_PART_TYPES_FOLDER).join(format!("{}.yaml", name));
        let yaml = std::fs::read_to_string(&file).expect(
            &format!(
                "Failed to read {}",
                file.to_str().map_or("invalid filename", |s| s)
            )[..],
        );
        serde_yaml::from_str(&yaml).map_err(|e| YamlError::from(e, file.to_path_buf()))
    }
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        serde_yaml::to_string(self)
    }
}

impl CustomPartTypeDefinition {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        CustomPartTypeDefinitionInput::from_normal(self.to_owned()).to_yaml()
    }
}

macro_rules! create_input_option_value {
    ($struct: ident, $input: literal, $type: ty, $numbas_subtype: ty) => {
        #[derive(Input, Overwrite, RumbasCheck)]
        #[input(name = $input)]
        #[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
        pub struct $struct {
            /// The value
            value: $type,
            /// A static field takes the same value in every instance of the part type. A dynamic field is defined by a JME expression which is evaluated when the question is run.
            #[serde(rename = "static")]
            is_static: bool,
        }

        impl
            ToNumbas<
                numbas::question::custom_part_type::CustomPartInputOptionValue<$numbas_subtype>,
            > for $struct
        {
            fn to_numbas(
                &self,
                locale: &str,
            ) -> numbas::question::custom_part_type::CustomPartInputOptionValue<$numbas_subtype>
            {
                numbas::question::custom_part_type::CustomPartInputOptionValue {
                    value: self.value.clone().to_numbas(locale),
                    is_static: self.is_static,
                }
            }
        }

        impl ToRumbas<$struct>
            for numbas::question::custom_part_type::CustomPartInputOptionValue<$numbas_subtype>
        {
            fn to_rumbas(&self) -> $struct {
                $struct {
                    value: self.value.clone().to_rumbas(),
                    is_static: self.is_static,
                }
            }
        }
    };
}

create_input_option_value!(
    CustomPartInputOptionValueBool,
    "CustomPartInputOptionValueBoolInput",
    bool,
    bool
);

create_input_option_value!(
    CustomPartInputOptionValueTranslatableString,
    "CustomPartInputOptionValueTranslatableStringInput",
    TranslatableString,
    String
);
create_input_option_value!(
    CustomPartInputOptionValueTranslatableStrings,
    "CustomPartInputOptionValueTranslatableStringsInput",
    TranslatableStrings,
    Vec<String>
);

create_input_option_value!(
    CustomPartInputOptionValueAnswerStyles,
    "CustomPartInputOptionValueAnswerStylesInput",
    Vec<crate::question::part::number_entry::AnswerStyle>,
    Vec<numbas::support::answer_style::AnswerStyle>
);

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "CustomPartInputWidgetInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(tag = "type")]
pub enum CustomPartInputWidget {
    //TODO other types: https://numbas-editor.readthedocs.io/en/latest/custom-part-types/reference.html
    #[serde(rename = "string")]
    /// The student enters a single line of text.
    String(CustomPartStringInputOptions),
    #[serde(rename = "number")]
    /// The student enters a number, using any of the allowed notation styles. If the studentâ€™s answer is not a valid number, they are shown a warning and can not submit the part.
    Number(CustomPartNumberInputOptions),
    #[serde(rename = "radiogroup")]
    /// The student chooses one from a list of choices by selecting a radio button.
    RadioGroup(CustomPartRadioGroupInputOptions),
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartInputWidget> for CustomPartInputWidget {
    fn to_numbas(&self, locale: &str) -> numbas::question::custom_part_type::CustomPartInputWidget {
        match self {
            CustomPartInputWidget::String(s) => {
                numbas::question::custom_part_type::CustomPartInputWidget::String(
                    s.to_numbas(locale),
                )
            }
            CustomPartInputWidget::Number(s) => {
                numbas::question::custom_part_type::CustomPartInputWidget::Number(
                    s.to_numbas(locale),
                )
            }
            CustomPartInputWidget::RadioGroup(s) => {
                numbas::question::custom_part_type::CustomPartInputWidget::RadioButtons(
                    s.to_numbas(locale),
                )
            }
        }
    }
}

impl ToRumbas<CustomPartInputWidget> for numbas::question::custom_part_type::CustomPartInputWidget {
    fn to_rumbas(&self) -> CustomPartInputWidget {
        match self {
            numbas::question::custom_part_type::CustomPartInputWidget::String(s) => {
                CustomPartInputWidget::String(s.to_rumbas())
            }
            numbas::question::custom_part_type::CustomPartInputWidget::Number(s) => {
                CustomPartInputWidget::Number(s.to_rumbas())
            }
            numbas::question::custom_part_type::CustomPartInputWidget::RadioButtons(s) => {
                CustomPartInputWidget::RadioGroup(s.to_rumbas())
            }
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "CustomPartStringInputOptionsInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
pub struct CustomPartStringInputOptions {
    //TODO? hint & correctAnswer is shared for all..., use macro?
    /// A string displayed next to the input field, giving any necessary information about how to enter their answer.
    pub hint: CustomPartInputOptionValueTranslatableString,
    /// A JME expression which evaluates to the expected answer to the part.
    pub correct_answer: JMETranslatableString,
    /// If false, the part will only be marked if their answer is non-empty.
    pub allow_empty: CustomPartInputOptionValueBool,
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartStringInputOptions>
    for CustomPartStringInputOptions
{
    fn to_numbas(
        &self,
        locale: &str,
    ) -> numbas::question::custom_part_type::CustomPartStringInputOptions {
        numbas::question::custom_part_type::CustomPartStringInputOptions {
            hint: self.hint.to_numbas(locale),
            correct_answer: self
                .correct_answer
                .to_string(locale)
                .unwrap()
                .try_into()
                .unwrap(),
            allow_empty: self.allow_empty.to_numbas(locale),
        }
    }
}

impl ToRumbas<CustomPartStringInputOptions>
    for numbas::question::custom_part_type::CustomPartStringInputOptions
{
    fn to_rumbas(&self) -> CustomPartStringInputOptions {
        CustomPartStringInputOptions {
            hint: self.hint.to_rumbas(),
            correct_answer: self.correct_answer.to_rumbas(),
            allow_empty: self.allow_empty.to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "CustomPartNumberInputOptionsInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
pub struct CustomPartNumberInputOptions {
    /// A string displayed next to the input field, giving any necessary information about how to enter their answer.
    pub hint: CustomPartInputOptionValueTranslatableString,
    /// A JME expression which evaluates to the expected answer to the part.
    pub correct_answer: JMETranslatableString,
    ///Allow the student to enter their answer as a fraction?
    pub allow_fractions: CustomPartInputOptionValueBool,
    pub allowed_notation_styles: CustomPartInputOptionValueAnswerStyles,
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartNumberInputOptions>
    for CustomPartNumberInputOptions
{
    fn to_numbas(
        &self,
        locale: &str,
    ) -> numbas::question::custom_part_type::CustomPartNumberInputOptions {
        numbas::question::custom_part_type::CustomPartNumberInputOptions {
            hint: self.hint.to_numbas(locale),
            correct_answer: self
                .correct_answer
                .to_string(locale)
                .unwrap()
                .try_into()
                .unwrap(),
            allow_fractions: self.allow_fractions.to_numbas(locale),
            allowed_notation_styles: self.allowed_notation_styles.to_numbas(locale),
        }
    }
}

impl ToRumbas<CustomPartNumberInputOptions>
    for numbas::question::custom_part_type::CustomPartNumberInputOptions
{
    fn to_rumbas(&self) -> CustomPartNumberInputOptions {
        CustomPartNumberInputOptions {
            hint: self.hint.to_rumbas(),
            correct_answer: self.correct_answer.to_rumbas(),
            allow_fractions: self.allow_fractions.to_rumbas(),
            allowed_notation_styles: self.allowed_notation_styles.to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "CustomPartRadioGroupInputOptionsInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
pub struct CustomPartRadioGroupInputOptions {
    /// A string displayed next to the input field, giving any necessary information about how to enter their answer.
    pub hint: CustomPartInputOptionValueTranslatableString,
    /// A JME expression which evaluates to the expected answer to the part.
    pub correct_answer: JMETranslatableString,
    /// The labels for the choices to offer to the student.
    pub choices: CustomPartInputOptionValueTranslatableStrings,
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartRadioButtonsInputOptions>
    for CustomPartRadioGroupInputOptions
{
    fn to_numbas(
        &self,
        locale: &str,
    ) -> numbas::question::custom_part_type::CustomPartRadioButtonsInputOptions {
        numbas::question::custom_part_type::CustomPartRadioButtonsInputOptions {
            hint: self.hint.to_numbas(locale),
            correct_answer: self
                .correct_answer
                .to_string(locale)
                .unwrap()
                .try_into()
                .unwrap(),
            choices: self.choices.to_numbas(locale),
        }
    }
}

impl ToRumbas<CustomPartRadioGroupInputOptions>
    for numbas::question::custom_part_type::CustomPartRadioButtonsInputOptions
{
    fn to_rumbas(&self) -> CustomPartRadioGroupInputOptions {
        CustomPartRadioGroupInputOptions {
            hint: self.hint.to_rumbas(),
            correct_answer: self.correct_answer.to_rumbas(),
            choices: self.choices.to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck, JsonSchema)]
#[input(name = "CustomPartTypeDefinitionPathInput")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(try_from = "String")]
#[serde(into = "String")]
pub struct CustomPartTypeDefinitionPath {
    pub custom_part_type_name: String,
    pub custom_part_type_data: CustomPartTypeDefinition,
}

impl ToNumbas<numbas::question::custom_part_type::CustomPartType> for CustomPartTypeDefinitionPath {
    fn to_numbas(&self, locale: &str) -> numbas::question::custom_part_type::CustomPartType {
        self.custom_part_type_data
            .clone()
            .to_numbas_with_name(locale, self.custom_part_type_name.clone())
    }
}

impl ToRumbas<CustomPartTypeDefinitionPath> for numbas::question::custom_part_type::CustomPartType {
    fn to_rumbas(&self) -> CustomPartTypeDefinitionPath {
        CustomPartTypeDefinitionPath {
            custom_part_type_data: CustomPartTypeDefinition {
                type_name: self.name.to_rumbas(),
                description: self.description.to_rumbas(),
                settings: self.settings.clone(),
                help_url: self.help_url.to_rumbas(),
                // public_availability: numbas::question::custom_part_type::CustomPartAvailability::Always,
                can_be_gap: self.can_be_gap,
                can_be_step: self.can_be_step,
                marking_notes: JMENotes(self.marking_notes.clone().to_rumbas()),
                published: self.published,
                extensions: Extensions::from(&self.extensions),
                input_widget: self.input_widget.to_rumbas(),
            },
            custom_part_type_name: self.short_name.clone(),
        }
    }
}

impl JsonSchema for CustomPartTypeDefinitionPathInput {
    fn schema_name() -> String {
        "CustomPartTypeDefinitionPath".to_owned()
    }

    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        gen.subschema_for::<String>()
    }
}

impl std::convert::TryFrom<String> for CustomPartTypeDefinitionPathInput {
    type Error = YamlError;

    fn try_from(s: String) -> Result<Self, Self::Error> {
        let custom_part_type_data = CustomPartTypeDefinitionInput::from_name(&s).map_err(|e| e)?;
        Ok(CustomPartTypeDefinitionPathInput {
            custom_part_type_name: Value::Normal(s),
            custom_part_type_data: Value::Normal(custom_part_type_data),
        })
    }
}

impl std::convert::From<CustomPartTypeDefinitionPathInput> for String {
    fn from(q: CustomPartTypeDefinitionPathInput) -> Self {
        q.custom_part_type_name.unwrap()
    }
}

// Remove these impl's etc, should not ser / deser
impl std::convert::TryFrom<String> for CustomPartTypeDefinitionPath {
    type Error = YamlError;

    fn try_from(s: String) -> Result<Self, Self::Error> {
        let data: CustomPartTypeDefinitionPathInput = s.try_into()?;
        Ok(data.to_normal())
    }
}

// Remove these impl's etc, should not ser / deser
impl std::convert::From<CustomPartTypeDefinitionPath> for String {
    fn from(q: CustomPartTypeDefinitionPath) -> Self {
        q.custom_part_type_name
    }
}

impl Hash for CustomPartTypeDefinitionPath {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.custom_part_type_name.hash(state);
    }
}
impl PartialEq for CustomPartTypeDefinitionPath {
    fn eq(&self, other: &Self) -> bool {
        self.custom_part_type_name == other.custom_part_type_name
    }
}
impl Eq for CustomPartTypeDefinitionPath {}
