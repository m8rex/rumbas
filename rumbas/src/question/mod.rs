//! Contains all the data types

pub mod constants;
pub mod custom_part_type;
pub mod extension;
pub mod feedback;
pub mod function;
pub mod navigation;
pub mod part;
pub mod preamble;
pub mod resource;
pub mod variable;
pub mod variable_test;

use crate::support::optional_overwrite::*;
use crate::support::template::TemplateData;
use crate::support::template::{Value, ValueType};
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::ContentAreaTranslatableString;
use crate::support::translatable::TranslatableString;
use crate::support::yaml::{YamlError, YamlResult};
use constants::BuiltinConstants;
use constants::CustomConstant;
use custom_part_type::CustomPartTypeDefinitionPath;
use extension::Extensions;
use function::Function;
use navigation::QuestionNavigation;
use part::question_part::QuestionPart;
use preamble::Preamble;
use resource::ResourcePath;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use variable::VariableRepresentation;
use variable::UNGROUPED_GROUP;
use variable_test::VariablesTest;

optional_overwrite! {
    pub struct Question {
        /// The statement is a content area which appears at the top of the question, before any input boxes. Use the statement to set up the question and provide any information the student needs to answer it.
        statement: ContentAreaTranslatableString,
        /// Advice is a content area which is shown when the student presses the Reveal button to reveal the questionâ€™s answers, or at the end of the exam.
        /// The advice area is normally used to present a worked solution to the question.
        advice: ContentAreaTranslatableString,
        parts: Vec<Value<QuestionPart>>,
        builtin_constants: BuiltinConstants,
        custom_constants: Vec<CustomConstant>,
        variables: HashMap<String, Value<VariableRepresentation>>,
        variables_test: VariablesTest,
        functions: HashMap<String, Value<Function>>,
        preamble: Preamble,
        navigation: QuestionNavigation,
        extensions: Extensions,
        /// The names of the topics used in diagnostic exams that this question belongs to
        diagnostic_topic_names: Vec<TranslatableString>, // TODO: validate? / warnings?
        resources: Vec<Value<ResourcePath>>,
        /// The custom part types used in this exam
        custom_part_types: Vec<CustomPartTypeDefinitionPath>
        //TODO a lot of options
    }
}

impl ToNumbas<numbas::exam::ExamQuestion> for Question {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::ExamQuestion {
        //TODO?
        panic!(
            "{}",
            "Should not happen, don't call this method Missing name".to_string(),
        )
    }
    //TODO: add to_numbas on Option's to reduce burden?
    fn to_numbas_with_name(&self, locale: &str, name: String) -> numbas::exam::ExamQuestion {
        if self.variables.unwrap().contains_key("e") {
            panic!("e is not allowed as a variable name"); //TODO is this still the case?
        }
        numbas::exam::ExamQuestion {
            name,
            statement: self.statement.to_numbas(locale),
            advice: self.advice.to_numbas(locale),
            parts: self.parts.to_numbas(locale),
            builtin_constants: numbas::exam::BuiltinConstants(
                self.builtin_constants.clone().unwrap().to_numbas(locale),
            ),
            constants: self.custom_constants.to_numbas(locale),
            variables: self
                .variables
                .clone()
                .unwrap()
                .into_iter()
                .map(|(k, v)| (k.clone(), v.to_numbas_with_name(locale, k)))
                .collect(),
            variables_test: self.variables_test.clone().unwrap().to_numbas(locale),
            functions: self.functions.to_numbas(locale),
            ungrouped_variables: self
                .variables
                .clone()
                .unwrap()
                .into_iter()
                .filter(|(_k, v)| &v.unwrap().to_variable().group.unwrap()[..] == UNGROUPED_GROUP)
                .map(|(k, _)| k)
                .collect(),
            variable_groups: Vec::new(), // Don't add variable groups
            rulesets: HashMap::new(),    //TODO: add to Question type ?
            preamble: self.preamble.clone().unwrap().to_numbas(locale),
            navigation: self.navigation.clone().unwrap().to_numbas(locale),
            extensions: self.extensions.clone().unwrap().to_numbas(locale),
            tags: self
                .diagnostic_topic_names
                .clone()
                .unwrap()
                .into_iter()
                .map(|t| format!("skill: {}", t.to_string(locale).unwrap()))
                .collect(),
            resources: self.resources.to_numbas(locale),
            custom_part_types: self
                .custom_part_types
                .clone()
                .unwrap()
                .into_iter()
                .map(|c| {
                    c.custom_part_type_data
                        .to_numbas_with_name(locale, c.custom_part_type_name)
                })
                .collect(),
        }
    }
}

impl ToRumbas<Question> for numbas::exam::ExamQuestion {
    fn to_rumbas(&self) -> Question {
        Question {
            statement: self.statement.to_rumbas(),
            advice: self.advice.to_rumbas(),
            parts: self.parts.to_rumbas(),
            builtin_constants: self.builtin_constants.to_rumbas(),
            custom_constants: self.constants.to_rumbas(),
            variables: self.variables.to_rumbas(),
            variables_test: self.variables_test.to_rumbas(),
            functions: self.functions.to_rumbas(),
            preamble: self.preamble.to_rumbas(),
            navigation: self.navigation.to_rumbas(),
            extensions: self.extensions.to_rumbas(),
            diagnostic_topic_names: Value::Normal(
                self.tags
                    .iter()
                    .filter(|t| t.starts_with("skill: "))
                    .map(|t| t.splitn(2, ": ").collect::<Vec<_>>()[1].to_string().into())
                    .collect(),
            ),
            resources: self.resources.to_rumbas(),
            custom_part_types: self.custom_part_types.to_rumbas(),
        }
    }
}

impl Question {
    pub fn from_name(name: &str) -> YamlResult<Question> {
        use QuestionFileType::*;
        let file = Path::new(crate::QUESTIONS_FOLDER).join(format!("{}.yaml", name));
        let yaml = fs::read_to_string(&file).expect(
            &format!(
                "Failed to read {}",
                file.to_str().map_or("invalid filename", |s| s)
            )[..],
        );
        let input: std::result::Result<QuestionFileType, serde_yaml::Error> =
            serde_yaml::from_str(&yaml);
        input
            .map(|e| match e {
                Normal(e) => Ok(*e),
                Template(t) => {
                    let template_file = Path::new(crate::QUESTION_TEMPLATES_FOLDER)
                        .join(format!("{}.yaml", t.relative_template_path));
                    let template_yaml = fs::read_to_string(&template_file).expect(
                        &format!(
                            "Failed to read {}",
                            template_file.to_str().map_or("invalid filename", |s| s)
                        )[..],
                    );
                    let mut question: Question = serde_yaml::from_str(&template_yaml).unwrap();
                    t.data.iter().for_each(|(k, v)| {
                        question.insert_template_value(k, &v.0);
                    });
                    Ok(question)
                }
            })
            .and_then(std::convert::identity) //flatten result is currently only possible in nightly
            .map_err(|e| YamlError::from(e, file.to_path_buf()))
    }
}

#[derive(Serialize, Deserialize, JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum QuestionFileType {
    Template(TemplateData),
    Normal(Box<Question>),
}

impl QuestionFileType {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        serde_yaml::to_string(self)
    }
}
