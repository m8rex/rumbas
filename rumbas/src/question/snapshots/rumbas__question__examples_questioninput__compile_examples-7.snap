---
source: rumbas/src/question/mod.rs
expression: "&example"
---
statement:
  template_key: template_key
  default_value: ~
advice:
  template_key: template_key
  default_value: ~
parts:
  template_key: template_key
  default_value: ~
builtin_constants: ~
custom_constants:
  template_key: template_key
  default_value: ~
variables:
  nonjmetext§: "10 .. 1#5"
variables_test: ~
functions:
  nonjmetext§:
    definition:
      template_key: template_key
      default_value: ~
    language: jme
    output_type: key_pair
    parameters:
      template_key: template_key
      default_value: ~
preamble: ~
navigation: ~
extensions: ~
diagnostic_topic_names:
  template_key: template_key
  default_value: ~
resources:
  template_key: template_key
  default_value: ~
custom_part_types: ~
rulesets:
  nonjmetext§: "template:template_key"

